### PR TITLE
JDK-8221642: AccessibleObject::setAccessible throws NPE when invoked by JNI code with no java frame on stack

### PR DESCRIPTION
**Pull request by mlchung**
`AccessibleObject::setAccessible` and `trySetAccessible` methods should only allow access to public member of a public type that is unconditionally exported consistent with the access check as described in the class specification, when invoked by JNI code with no Java class on the stack.   The current implementation throws NPE when finding the module of the caller class as the caller class is null.

The specification of `canAccess`, `setAccessible` and `trySetAccessible` are updated to specify the behavior when the caller class is null.   I consider this spec update as a clarification as the class specification covers this case.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed
- [x] Change requires a CSR request to be approved

### Issues
 * [JDK-8221642](https://bugs.openjdk.java.net/browse/JDK-8221642): AccessibleObject::setAccessible throws NPE when invoked by JNI code with no java frame on stack
 * [JDK-8280831](https://bugs.openjdk.java.net/browse/JDK-8280831): AccessibleObject::setAccessible throws NPE when invoked by JNI code with no java frame on stack (**CSR**)

### Reviewers
 * [Alan Bateman](https://openjdk.java.net/census#alanb) (@AlanBateman - **Reviewer**) ⚠️ Review applies to 0f57385db06869ffed308b176b99ae77ca57c144

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/7271/head:pull/7271` \
`$ git checkout pull/7271`

Update a local copy of the PR: \
`$ git checkout pull/7271` \
`$ git pull https://git.openjdk.java.net/jdk pull/7271/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 7271`

View PR using the GUI difftool: \
`$ git pr show -t 7271`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/7271.diff">https://git.openjdk.java.net/jdk/pull/7271.diff</a>

</details>

### CODE REVIEW

**Review by AlanBateman — APPROVED** [review]

File: src/java.base/share/classes/java/lang/reflect/AccessibleObject.java [hunk 1]
```diff
@@ -246,6 +255,11 @@ boolean setAccessible0(boolean flag) {
      *     }
      * }</pre>
      *
+     * <p> If this method is invoked by <a href="{@docRoot}/../specs/jni/index.html">JNI code</a>
+     * with no caller class on the stack, the {@code accessible} flag can
+     * only set if the member and the declaring class are public, and
```
Comment (dholmes-ora) [inline src/java.base/share/classes/java/lang/reflect/AccessibleObject.java:260]: s/only set/only be set/

**Comment by dholmes-ora** [comment]
Hi Mandy,

There is no spec change here as you note, but the behavioural change also makes this warrant a CSR request IMO.

Cheers,
David

**Comment by AlanBateman** [comment]
> There is no spec change here as you note, but the behavioural change also makes this warrant a CSR request IMO.

It's in draft: https://bugs.openjdk.java.net/browse/JDK-8280831, I assume Mandy will finalize shortly.

**Comment by dholmes-ora** [comment]
Apologies as I somehow missed the CSR info above.

File: src/java.base/share/classes/java/lang/reflect/AccessibleObject.java (hunk 1 above)
Comment (mlchung) [reply src/java.base/share/classes/java/lang/reflect/AccessibleObject.java]: Thanks for catching this. Fixed

**Comment by mlchung** [comment]
Right, I have the CSR to cover the behavioral change.    I called out the spec change in this patch is a spec clarification as this fix can be considered to backport without the spec change.

**Pull request closed without merging**